From: "Matthew Wilcox (Oracle)" <willy@infradead.org>
Subject: mm: convert destroy_compound_page() to destroy_large_folio()
Date: Fri, 17 Jun 2022 18:50:17 +0100

All callers now have a folio, so push the folio->page conversion
down to this function.

Link: https://lkml.kernel.org/r/20220617175020.717127-20-willy@infradead.org
Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/mm.h |    8 +++++---
 mm/swap.c          |    2 +-
 mm/vmscan.c        |    4 ++--
 3 files changed, 8 insertions(+), 6 deletions(-)

--- a/include/linux/mm.h~mm-convert-destroy_compound_page-to-destroy_large_folio
+++ a/include/linux/mm.h
@@ -923,10 +923,12 @@ static inline void set_compound_page_dto
 	page[1].compound_dtor = compound_dtor;
 }
 
-static inline void destroy_compound_page(struct page *page)
+static inline void destroy_large_folio(struct folio *folio)
 {
-	VM_BUG_ON_PAGE(page[1].compound_dtor >= NR_COMPOUND_DTORS, page);
-	compound_page_dtors[page[1].compound_dtor](page);
+	enum compound_dtor_id dtor = folio_page(folio, 1)->compound_dtor;
+
+	VM_BUG_ON_FOLIO(dtor >= NR_COMPOUND_DTORS, folio);
+	compound_page_dtors[dtor](&folio->page);
 }
 
 static inline int head_compound_pincount(struct page *head)
--- a/mm/swap.c~mm-convert-destroy_compound_page-to-destroy_large_folio
+++ a/mm/swap.c
@@ -115,7 +115,7 @@ static void __folio_put_large(struct fol
 	 */
 	if (!folio_test_hugetlb(folio))
 		__page_cache_release(folio);
-	destroy_compound_page(&folio->page);
+	destroy_large_folio(folio);
 }
 
 void __folio_put(struct folio *folio)
--- a/mm/vmscan.c~mm-convert-destroy_compound_page-to-destroy_large_folio
+++ a/mm/vmscan.c
@@ -1985,7 +1985,7 @@ free_it:
 		 * appear not as the counts should be low
 		 */
 		if (unlikely(folio_test_large(folio)))
-			destroy_compound_page(&folio->page);
+			destroy_large_folio(folio);
 		else
 			list_add(&folio->lru, &free_pages);
 		continue;
@@ -2354,7 +2354,7 @@ static unsigned int move_pages_to_lru(st
 
 			if (unlikely(folio_test_large(folio))) {
 				spin_unlock_irq(&lruvec->lru_lock);
-				destroy_compound_page(&folio->page);
+				destroy_large_folio(folio);
 				spin_lock_irq(&lruvec->lru_lock);
 			} else
 				list_add(&folio->lru, &folios_to_free);
_
